syntax = "proto3";

//import "llm.proto";
import "core.proto";
import "google/protobuf/any.proto";
import "google/protobuf/struct.proto";

package instinct.retrieval;


message BoolQuery {
  repeated SearchQuery must = 1;
  repeated SearchQuery should = 2;
  repeated SearchQuery mustNot = 3;
}

message TermQuery {
  google.protobuf.Value term = 1;
  string name = 2;
}

message TermsQuery {
  repeated google.protobuf.Value terms = 1;
  string name = 2;
}


message SearchQuery {
  oneof query {
    BoolQuery bool = 1;
    TermQuery term = 2;
  }
}

message SearchRequest {
  string query = 1;
  int32 top_k = 2;
  SearchQuery metadata_filter = 3;
}

message SearchResponse {

}

message UpdateResult {
  int32 affected_rows = 1;
  repeated string returned_ids = 2;
  repeated instinct.core.Document failed_documents = 3;
}

message VectorStoreInstanceMetadata {
  string instance_id = 1;
  core.MetadataSchema metadata_schema = 2;
  int64 created_at = 3;
  int64 modified_at = 4;
  google.protobuf.Any custom = 100;
}
